source/__tests__/integration/new-profile.test.js

```diff
@@ -16,166 +17,216 @@ describe('Adding and removing a new Profile', () => {
+      const resourceSelectorSel = `${resourceChooseModalSel} > select#resourcePick`
+      await expect(page).toSelect(resourceSelectorSel, 'Role')
 
-    it('now has no span or input fields for the resource template', async () => {
-      await expect(page).not.toMatch('Property Template')
-      await expect(page).not.toMatch('span', {
-        text: 'Resource Template'
-      })
+      await expect_regex_in_sel_textContent('span[popover="URI: http://id.loc.gov/ontologies/bibframe/Role"]', 'Role')
     })
-  })
 
-  describe('Profile fields errors', () => {
+    it('removing resource template removes its span and input fields', async () => {
+      expect.assertions(4)
+
+      const deleteTargetSel = 'span[popover="URI: http://id.loc.gov/ontologies/bibframe/Role"]'
+      await page.waitForSelector(deleteTargetSel)
+      await expect_regex_in_sel_textContent(deleteTargetSel, 'Role')
 
-    it('displays error if source and user switches focus', async () => {
-      await page
-        .type(source_input_sel, '')
-        .catch(error => console.log(`promise error for : ${error}`))
-      await page.keyboard.press(`Tab`)
-      await expect_value_in_selector_textContent('#alertBox > div > div > div.modal-header > h3', 'Error!')
-      const invalid_url_class = page.$('input#resourceURI', e => e.getAttribute('ng-invalid-url'))
-      expect(invalid_url_class).toBeDefined()
+      const deleteRoleSel = 'div[name="resourceForm"] div#resource_0 #deleteButton'
+      await page.waitForSelector(deleteRoleSel, {visible: true})
+      await page.waitFor(500) // shameless green: it needs more time here; can't tell why
+      await expect(page).toClick(deleteRoleSel)
+
+      const modalConfirmDeleteSel = 'div#deleteModal div.modal-body > button[ng-click="confirm();"]'
+      await page.waitForSelector(modalConfirmDeleteSel, {visible: true})
+      await expect(page).toClick(modalConfirmDeleteSel)
+
+      await expect_sel_not_to_exist(deleteTargetSel)
     })
-  })
 
+    it('alerts with error if non-uri profile source is entered and user switches focus', async () => {
+      expect.assertions(4)
+      await page.goto('http://localhost:8000/#/profile/create/')
+      await expect(page).toFill(sourceInputSel, 'nope')
+      await expect(page).toClick('#profile input[name="description"]')
+      const alertBoxSel = '#alertBox > div > div > div.modal-header > h3'
+      await page.waitForSelector(alertBoxSel, {visible: true})
+      await expect_value_in_selector_textContent(alertBoxSel, 'Error!')
+      await page.waitForSelector('#alertClose', {visible: true})
+      await page.click('#alertClose')
+      const invalid_url_class = await page.$(sourceInputSel, e => e.getAttribute('ng-invalid-url'))
+      expect(invalid_url_class).toBeTruthy()
+    })
+  })
 })
 
+async function expect_regex_in_sel_textContent(sel, value) {
+  const sel_text = await page.$eval(sel, e => e.textContent)
+  expect(sel_text).toMatch(value)
```

Comment:
No `return` needed here?

Reply:
whoops!  my bad!

Reply:
wait -- these are, effectively, "expect(x)" statements;  they do not return promises, and they are accounted for with `expect.assertions(n)` in the tests.

Reply:
wow - my ignorance knows no bounds.  So all async functions return a promise.  return is more correct, yes.

Reply:
actually, this seems fine to me.  `sel_text` should indeed just be text, and thus the expectation on it (AFAIK) should be regular synchronous code.

so, i actually somewhat doubt that `return` is more correct, though i do agree (based on my testing and the docs) that all `async` functions return a `Promise`.  i think that's actually why the `return` in a `return await ...` might be superfluous.

i think unless you care about the value returned by the `expect` (as opposed to just caring that the `expect` executes), all that matters is that the test waits on completion of the method that contains the expectation.  i think `await`ing on the `async` methods (such as this one) accomplishes that goal.

i was feeling pretty confused by all of this, and so i wrote some test code that tries to illustrate some of the nuances of `async`/`await` behavior, so that i could better grok it.  i think/hope i have a better less confused grasp on it.  i'll post my test code on this PR, since there's discussion happening here around this, but happy to move it to a gist and discussion in channel or something.

if my conclusions based on my test code are subtly or totally wrong, i'd obviously love to know that too.  so i'm interested to see whether people come to the same conclusions i did.

Reply:
FWIW, changing all of these methods to use "return" *seemed* to make for more consistent results.  But that is completely anecdotal and has confirmation bias.